skbio/tree/tests/test_nj.py

```diff
@@ -191,6 +221,201 @@ def test_pair_members_to_new_node_zero_branch_length(self):
         self.assertEqual(
             _pair_members_to_new_node(dm, 'a', 'b', False), (-16, 20))
 
+    def test_nni_dm1(self):
+        self.assertEqual(nj(self.dm1, result_constructor=str),
+                         self.expected1_str)
+        actual_TreeNode = nni(self.pre1_nni_TreeNode, self.dm1, inplace=False)
+        self.assertAlmostEqual(actual_TreeNode.compare_tip_distances(
+            self.post1_nni_TreeNode), 0.0, places=10)
+
+    def test_nni_dm2(self):
+        # Resulting tree topology is equivalent to result from nj, however,
+        # resulting edge lengths are almost equal to 2 places.
+        actual_TreeNode = nni(self.pre2_nni_TreeNode, self.dm2, inplace=False)
+        self.assertAlmostEqual(actual_TreeNode.compare_tip_distances(
+            self.post2_nni_TreeNode), 0.0)
+
+    def test_nni_dm3(self):
+        actual_TreeNode = nni(self.pre3_nni_TreeNode, self.dm3, inplace=False)
+        self.assertAlmostEqual(actual_TreeNode.compare_tip_distances(
+            self.post3_nni_TreeNode), 0.0)
+        
+    def test_nni_trivial(self):
+        # No swaps are performed, but edge lengths are assigned.
+        data = [[0, 3, 2],
+                [3, 0, 3],
+                [2, 3, 0]]
+        dm = DistanceMatrix(data, list('abc'))
+        pre_str = "((c,b))a;"
+        pre_TreeNode = TreeNode.read(
+                io.StringIO(pre_str))
+        expected_str = "((c:1.0,b:2.0):1.0)a;"
+        expected_TreeNode = TreeNode.read(
+                io.StringIO(expected_str))
+        self.assertEqual(str(nni(pre_TreeNode, dm, inplace=False)),
+                         str(expected_TreeNode))
+
+    def test_nni_binary_flag(self):
```

Comment:
The way to test errors in Python is typically to use a [context manager](https://book.pythontips.com/en/latest/context_managers.html). So here it would look something like:

````python
with self.assertRaises(TypeError, msg="Could not perform NNI. Tree needs to be a binary tree."):
        data = [[0, 3],
                [3, 0]]
        dm = DistanceMatrix(data, list('ab'))
        pre_str = "((b))a;"
        pre_TreeNode = TreeNode.read(
                io.StringIO(pre_str))
        nni(pre_TreeNode, dm)
````

Reply:
Thank you!

Reply:
@mataton @christapo Actually, this is probably not the very best example. Instead you may want to do:

```python
data = [[0, 3],
        [3, 0]]
dm = DistanceMatrix(data, list('ab'))
pre_str = "((b))a;"
pre_TreeNode = TreeNode.read(io.StringIO(pre_str))
msg = "Could not perform NNI. Tree needs to be a binary tree."
with self.assertRaisesRegex(TypeError, msg):
    nni(pre_TreeNode, dm)
```

Rationales: 1) The error testing context manager should contain only the code that will raise an error, that is, the `nni` call. The preparations don't have to be included. This way can precisely narrow down the test to the `nni` call.

2). The `msg` parameter of [`assertRaises`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaises) is very confusing. It doesn't actually check the error message. Therefore the original code won't work. Instead, [`assertRaisesRegex`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaisesRegex) will. [Here](https://stackoverflow.com/questions/58687475/) are some discussions.

If you do intend to use `assertRaises`, the correct code is:

```python
with self.assertRaises(TypeError) as cm:
    nni(pre_TreeNode, dm)
self.assertEqual(str(cm.exception), msg)
```

(I preferred this way because it is more precise than `assertRaisesRegex`.)

Reply:
Thank you for your insight! I've updated the error testing for both the binary tree test here and the leaf root test below.